File: 20-valid-parentheses/valid-parentheses.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()

        brackets = {")": "(", "}": "{", "]": "["}

        for p in s:
            if p in brackets.values():
                stack.append(p)
            elif stack and brackets[p] == stack[-1]:
                stack.pop()
            else:
                return False

        return len(stack) == 0

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<String> stack = new Stack<String>();

        HashMap<String, String> hashmap = new HashMap<>(Map.of(
                ")", "(",
                "}", "{",
                "]", "["));

        for (int p = 0; p < s.length(); p++) {
            String currentChar = Character.toString(s.charAt(p));

            if (hashmap.containsValue(currentChar)) {
                stack.push(currentChar);
            } else if (!stack.isEmpty() && hashmap.get(currentChar).equals(stack.peek())) {
                stack.pop();
            } else {
                return false;
            }
        }

        return stack.isEmpty();
    }

}
